File: src/main/java/org/gautelis/durga/plugins/JsonFlatten.java
package org.gautelis.durga.plugins;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.Iterator;
import java.util.Map;

/**
 * Flattens nested JSON to dot-notation keys, or unflattens dot-notation keys
 * back to nested objects.
 *
 * <p>Config syntax:
 * <pre>
 * "direction=flatten;separator=."   (flatten nested to top-level)
 * "direction=unflatten;separator=." (nest dot-notation keys)
 * "direction=flatten;separator=.;maxDepth=3"
 * </pre>
 */
public final class JsonFlatten implements Plugin {

    @Override
    public String execute(String payload, String config) throws Exception {
        String direction = "flatten";
        String separator = ".";
        int maxDepth = Integer.MAX_VALUE;
        if (config != null && !config.isBlank()) {
            String[] parts = config.split(";");
            for (String part : parts) {
                part = part.trim();
                int eq = part.indexOf('=');
                if (eq > 0) {
                    String key = part.substring(0, eq).trim();
                    String val = part.substring(eq + 1).trim();
                    switch (key) {
                        case "direction" -> direction = val;
                        case "separator" -> separator = val;
                        case "maxDepth" -> {
                            try { maxDepth = Integer.parseInt(val); } catch (NumberFormatException ignored) {}
                        }
                    }
                }
            }
        }
        if ("unflatten".equalsIgnoreCase(direction)) {
            return unflatten(payload, separator);
        }
        return flatten(payload, separator, maxDepth);
    }

    private JsonFlatten() {
    }

    public static String flatten(String json, String separator, int maxDepth) {
        ObjectMapper mapper = new ObjectMapper();
        JsonNode input;
        try {
            input = mapper.readTree(json);
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("Invalid input JSON", e);
        }
        if (!input.isObject()) {
            return json;
        }

        ObjectNode output = mapper.createObjectNode();
        flattenNode(mapper, output, "", (ObjectNode) input, separator, 0, maxDepth);
        return output.toString();
    }

    private static void flattenNode(ObjectMapper mapper, ObjectNode output, String prefix,
                                     ObjectNode node, String separator, int depth, int maxDepth) {
        Iterator<Map.Entry<String, JsonNode>> fields = node.fields();
        while (fields.hasNext()) {
            Map.Entry<String, JsonNode> entry = fields.next();
            String key = prefix.isEmpty() ? entry.getKey() : prefix + separator + entry.getKey();
            JsonNode value = entry.getValue();
            if (value.isObject() && !value.isEmpty() && depth < maxDepth) {
                flattenNode(mapper, output, key, (ObjectNode) value, separator, depth + 1, maxDepth);
            } else if (value.isArray()) {
                output.set(key, value);
            } else {
                output.set(key, value);
            }
        }
    }

    public static String unflatten(String json, String separator) {
        ObjectMapper mapper = new ObjectMapper();
        JsonNode input;
        try {
            input = mapper.readTree(json);
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("Invalid input JSON", e);
        }
        if (!input.isObject()) {
            return json;
        }

        ObjectNode output = mapper.createObjectNode();
        Iterator<Map.Entry<String, JsonNode>> fields = input.fields();
        while (fields.hasNext()) {
            Map.Entry<String, JsonNode> entry = fields.next();
            String key = entry.getKey();
            JsonNode value = entry.getValue();
            PipelinePlugin.setFieldAt(output, key, value);
        }
        return output.toString();
    }
}

File: durga-rust/src/plugins/json_flatten.rs
//! Port of `org.gautelis.durga.plugins.JsonFlatten` (plugin id `json-flatten`).
//!
//! `direction=flatten` collapses nested objects to `separator`-joined keys;
//! `direction=unflatten` nests `separator`-joined keys back into objects.
//! Arrays are kept as leaf values.

use serde_json::{Map, Value};

use crate::pipeline::set_field_at;
use crate::plugin::{Plugin, PluginError};

#[derive(Default)]
pub struct JsonFlatten;

impl JsonFlatten {
    pub fn new() -> Self {
        JsonFlatten
    }

    pub fn flatten(json: &str, separator: &str, max_depth: usize) -> Result<String, PluginError> {
        let input: Value = serde_json::from_str(json)
            .map_err(|e| -> PluginError { format!("Invalid input JSON: {e}").into() })?;
        let Value::Object(obj) = &input else {
            return Ok(json.to_string());
        };
        let mut output = Map::new();
        flatten_node(&mut output, "", obj, separator, 0, max_depth);
        Ok(Value::Object(output).to_string())
    }

    pub fn unflatten(json: &str, separator: &str) -> Result<String, PluginError> {
        let input: Value = serde_json::from_str(json)
            .map_err(|e| -> PluginError { format!("Invalid input JSON: {e}").into() })?;
        let Value::Object(obj) = &input else {
            return Ok(json.to_string());
        };
        let mut output = Value::Object(Map::new());
        for (key, value) in obj {
            set_field_at_sep(&mut output, key, separator, value.clone());
        }
        Ok(output.to_string())
    }
}

fn flatten_node(
    output: &mut Map<String, Value>,
    prefix: &str,
    node: &Map<String, Value>,
    separator: &str,
    depth: usize,
    max_depth: usize,
) {
    for (k, value) in node {
        let key = if prefix.is_empty() { k.clone() } else { format!("{prefix}{separator}{k}") };
        match value {
            Value::Object(child) if !child.is_empty() && depth < max_depth => {
                flatten_node(output, &key, child, separator, depth + 1, max_depth);
            }
            _ => {
                output.insert(key, value.clone());
            }
        }
    }
}

/// Sets a value at a separator-joined path. When the separator is `.` this is
/// equivalent to [`set_field_at`]; otherwise the path is split on the custom
/// separator.
fn set_field_at_sep(root: &mut Value, path: &str, separator: &str, value: Value) {
    if separator == "." {
        set_field_at(root, path, value);
        return;
    }
    let segments: Vec<&str> = path.split(separator).collect();
    if !root.is_object() {
        *root = Value::Object(Map::new());
    }
    let mut current = root;
    for segment in &segments[..segments.len() - 1] {
        let map = current.as_object_mut().expect("ensured object");
        let child = map.entry((*segment).to_string()).or_insert(Value::Null);
        if !child.is_object() {
            *child = Value::Object(Map::new());
        }
        current = child;
    }
    if let Value::Object(map) = current {
        map.insert(segments[segments.len() - 1].to_string(), value);
    }
}

impl Plugin for JsonFlatten {
    fn execute_str(&self, payload: &str, config: &str) -> Result<Option<String>, PluginError> {
        let mut direction = "flatten".to_string();
        let mut separator = ".".to_string();
        let mut max_depth = usize::MAX;
        for part in config.split(';') {
            let part = part.trim();
            let Some(eq) = part.find('=') else { continue };
            if eq == 0 {
                continue;
            }
            let key = part[..eq].trim();
            let val = part[eq + 1..].trim();
            match key {
                "direction" => direction = val.to_string(),
                "separator" => separator = val.to_string(),
                "maxDepth" => {
                    if let Ok(n) = val.parse::<usize>() {
                        max_depth = n;
                    }
                }
                _ => {}
            }
        }
        let out = if direction.eq_ignore_ascii_case("unflatten") {
            Self::unflatten(payload, &separator)?
        } else {
            Self::flatten(payload, &separator, max_depth)?
        };
        Ok(Some(out))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn flattens_nested() {
        let out = JsonFlatten::new()
            .execute_str(r#"{"a":{"b":{"c":1}},"d":2}"#, "direction=flatten;separator=.")
            .unwrap()
            .unwrap();
        let v: Value = serde_json::from_str(&out).unwrap();
        assert_eq!(v["a.b.c"], 1);
        assert_eq!(v["d"], 2);
    }

    #[test]
    fn unflattens_dot_keys() {
        let out = JsonFlatten::new()
            .execute_str(r#"{"a.b.c":1,"d":2}"#, "direction=unflatten;separator=.")
            .unwrap()
            .unwrap();
        let v: Value = serde_json::from_str(&out).unwrap();
        assert_eq!(v["a"]["b"]["c"], 1);
        assert_eq!(v["d"], 2);
    }
}
